File: app/rpc/open-in-app.ts
/** Open-in-app RPC — port of electron/ipc/openInApp.ts (tide:openInApp:detect
 *  / open). Detects external apps (Finder/Explorer, Terminal, VSCode, Zed)
 *  and opens a session's resolved folder in one. Electrobun substitution: the
 *  Electron version read OS icons via app.getFileIcon + sips; the devkit has
 *  no file-icon API, so iconDataUrl is always null and the renderer falls
 *  back to its lucide icons (a documented 4.x gap — the fallback path the
 *  Electron version already had for Linux). shell.openPath is replaced by
 *  Utils.openPath. */

import { spawn, spawnSync } from 'node:child_process';
import * as fs from 'node:fs';
import * as os from 'node:os';
import { Utils } from 'electrobun/main';
import type { ExternalApp, ExternalAppTarget, ShellOpResult } from '../../shared/rpc';

// Detection result cached for the process lifetime — installing an editor
// mid-session requires a restart to surface (same trade-off as the Electron
// shell).
let detectedCache: ExternalApp[] | null = null;

/** True if a CLI binary is on PATH (which/where) OR — macOS only — the .app
 *  bundle exists. */
function isEditorAvailable(cli: string, macBundle?: string): boolean {
  const probe = process.platform === 'win32'
    ? spawnSync('where', [cli], { stdio: 'ignore' })
    : spawnSync('which', [cli], { stdio: 'ignore' });
  if (probe.status === 0) return true;
  if (process.platform === 'darwin' && macBundle) {
    try {
      return fs.existsSync(`/Applications/${macBundle}`);
    } catch {
      /* ignore */
    }
  }
  return false;
}

/** The OS-appropriate display name for the built-in file manager. The target
 *  id stays 'finder' (a stable cross-platform identifier persisted in the
 *  renderer), but the label matches the OS. */
function fileManagerLabel(): string {
  if (process.platform === 'win32') return 'File Explorer';
  if (process.platform === 'darwin') return 'Finder';
  return 'Files';
}

/** Detect available apps. Icons are not extractable without a devkit
 *  file-icon API — always null (renderer falls back to lucide icons). */
function detectApps(): ExternalApp[] {
  if (detectedCache) return detectedCache;
  detectedCache = [
    { id: 'finder', label: fileManagerLabel(), available: true, iconDataUrl: null },
    { id: 'terminal', label: 'Terminal', available: true, iconDataUrl: null },
    {
      id: 'vscode',
      label: 'VSCode',
      available: isEditorAvailable('code', 'Visual Studio Code.app'),
      iconDataUrl: null,
    },
    {
      id: 'zed',
      label: 'Zed',
      available: isEditorAvailable('zed', 'Zed.app'),
      iconDataUrl: null,
    },
  ];
  return detectedCache;
}

/** Spawn a detached process (stdio:'ignore' + unref) so the app outlives Tide. */
function detach(cmd: string, args: string[], opts?: { cwd?: string; shell?: boolean }): boolean {
  try {
    const child = spawn(cmd, args, {
      detached: true,
      stdio: 'ignore',
      cwd: opts?.cwd ?? undefined,
      shell: opts?.shell ?? false,
    });
    child.unref();
    return true;
  } catch {
    return false;
  }
}

/** Is a CLI binary on PATH? (Unix `which` / Windows `where`.) */
function cliAvailable(cli: string): boolean {
  try {
    const r = process.platform === 'win32'
      ? spawnSync('where', [cli], { stdio: 'ignore' })
      : spawnSync('which', [cli], { stdio: 'ignore' });
    return r.status === 0;
  } catch {
    return false;
  }
}

/** Launch an editor for `dir`: prefer the CLI on PATH; fall back to
 *  `open -a <App>` on macOS for installed .apps missing the CLI. */
function launchEditor(cli: string, macAppName: string, dir: string): boolean {
  if (cliAvailable(cli)) {
    return detach(cli, [dir], { shell: process.platform === 'win32' });
  }
  if (process.platform === 'darwin') {
    return detach('open', ['-a', macAppName, dir]);
  }
  return false;
}

function openInTarget(target: ExternalAppTarget, dir: string): ShellOpResult {
  if (!fs.existsSync(dir)) {
    return { ok: false, error: `Path does not exist: ${dir}` };
  }
  switch (target) {
    case 'finder': {
      // Utils.openPath opens the directory in the OS file manager; it reports
      // success as a boolean rather than Electron's error-string promise.
      const opened = Utils.openPath(dir);
      return opened ? { ok: true } : { ok: false, error: 'Failed to open file manager' };
    }
    case 'terminal': {
      if (process.platform === 'darwin') {
        return detach('open', ['-a', 'Terminal', dir])
          ? { ok: true }
          : { ok: false, error: 'Failed to launch Terminal' };
      }
      if (process.platform === 'win32') {
        return detach('cmd', ['/c', 'start', '', 'cmd'], { cwd: dir })
          ? { ok: true }
          : { ok: false, error: 'Failed to launch cmd' };
      }
      return detach('x-terminal-emulator', ['--working-directory', dir]) ||
        detach('xdg-open', [dir])
        ? { ok: true }
        : { ok: false, error: 'No terminal handler found' };
    }
    case 'vscode':
      return launchEditor('code', 'Visual Studio Code', dir)
        ? { ok: true }
        : { ok: false, error: 'Failed to launch VSCode' };
    case 'zed':
      return launchEditor('zed', 'Zed', dir)
        ? { ok: true }
        : { ok: false, error: 'Failed to launch Zed' };
    default:
      return { ok: false, error: `Unknown target: ${target}` };
  }
}

export interface OpenInAppRpcOpts {
  /** Resolve a session's folder: worktree.path → workspace path → $HOME. */
  resolveSessionPath?: (sessionId?: string) => string;
}

export function registerOpenInAppRpc(opts: OpenInAppRpcOpts = {}) {
  const resolveSessionPath =
    opts.resolveSessionPath ??
    ((_sessionId?: string) => os.homedir());

  return {
    openInAppDetect: (_: Record<string, never>) => detectApps(),

    openInAppOpen: ({ target, sessionId }: { target: ExternalAppTarget; sessionId?: string }) => {
      const dir = resolveSessionPath(sessionId);
      return openInTarget(target, dir);
    },
  };
}

File: src-tauri/src/commands/open_in_app.rs
//! Open-in-app commands — port of `app/rpc/open-in-app.ts`: detects
//! external apps (Finder/Files, Terminal, VSCode, Zed) and opens a
//! session's resolved folder in one. Icons are not extractable here —
//! always null (the renderer falls back to its lucide icons).

use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};

use serde::Serialize;

use crate::state::AppState;

use super::CommandError;

/// `ExternalApp` (shared/rpc.ts).
#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct ExternalAppWire {
    pub id: &'static str,
    pub label: String,
    pub available: bool,
    #[serde(rename = "iconDataUrl")]
    pub icon_data_url: Option<String>,
}

/// `ShellOpResult`.
#[derive(Debug, Serialize, PartialEq)]
pub struct ShellOpResultWire {
    pub ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Detection cached for the process lifetime — installing an editor
/// mid-session requires a restart to surface.
fn detected_cache() -> &'static std::sync::Mutex<Option<Vec<ExternalAppWire>>> {
    static CACHE: std::sync::OnceLock<std::sync::Mutex<Option<Vec<ExternalAppWire>>>> =
        std::sync::OnceLock::new();
    CACHE.get_or_init(|| std::sync::Mutex::new(None))
}

/// True if a CLI binary is on PATH (which/where) OR — macOS only — the
/// .app bundle exists.
fn is_editor_available(cli: &str, mac_bundle: Option<&str>) -> bool {
    if cli_available(cli) {
        return true;
    }
    if cfg!(target_os = "macos") {
        if let Some(bundle) = mac_bundle {
            return Path::new("/Applications").join(bundle).exists();
        }
    }
    false
}

/// The OS-appropriate display name for the built-in file manager.
fn file_manager_label() -> &'static str {
    if cfg!(windows) {
        "File Explorer"
    } else if cfg!(target_os = "macos") {
        "Finder"
    } else {
        "Files"
    }
}

fn detect_apps() -> Vec<ExternalAppWire> {
    let mut cache = detected_cache().lock().expect("open-in-app cache poisoned");
    if let Some(cached) = cache.as_ref() {
        return cached.clone();
    }
    let apps = vec![
        ExternalAppWire {
            id: "finder",
            label: file_manager_label().to_string(),
            available: true,
            icon_data_url: None,
        },
        ExternalAppWire {
            id: "terminal",
            label: "Terminal".into(),
            available: true,
            icon_data_url: None,
        },
        ExternalAppWire {
            id: "vscode",
            label: "VSCode".into(),
            available: is_editor_available("code", Some("Visual Studio Code.app")),
            icon_data_url: None,
        },
        ExternalAppWire {
            id: "zed",
            label: "Zed".into(),
            available: is_editor_available("zed", Some("Zed.app")),
            icon_data_url: None,
        },
    ];
    *cache = Some(apps.clone());
    apps
}

/// Is a CLI binary on PATH? (Unix `which` / Windows `where`.)
fn cli_available(cli: &str) -> bool {
    let probe = if cfg!(windows) {
        Command::new("where")
            .args([cli])
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status()
    } else {
        Command::new("which")
            .args([cli])
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status()
    };
    probe.is_ok_and(|s| s.success())
}

/// Spawn a detached process so the launched app outlives Tide.
fn detach(cmd: &str, args: &[&str], cwd: Option<&Path>) -> bool {
    let mut command = Command::new(cmd);
    #[cfg(unix)]
    {
        use std::os::unix::process::CommandExt as _;
        command.process_group(0);
    }
    command
        .args(args)
        .current_dir(cwd.unwrap_or_else(|| Path::new("/")))
        .stdin(Stdio::null())
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .spawn()
        .is_ok()
}

/// Launch an editor for `dir`: prefer the CLI on PATH; fall back to
/// `open -a <App>` on macOS for installed .apps missing the CLI.
fn launch_editor(cli: &str, mac_app: &str, dir: &Path) -> bool {
    if cli_available(cli) {
        let dir = dir.to_string_lossy().into_owned();
        return detach(cli, &[&dir], None);
    }
    if cfg!(target_os = "macos") {
        let dir = dir.to_string_lossy().into_owned();
        return detach("open", &["-a", mac_app, &dir], None);
    }
    false
}

fn open_in_target(target: &str, dir: &Path) -> ShellOpResultWire {
    if !dir.exists() {
        return ShellOpResultWire {
            ok: false,
            error: Some(format!("Path does not exist: {}", dir.display())),
        };
    }
    let dir_str = dir.to_string_lossy().into_owned();
    match target {
        "finder" => {
            // `open` (macOS) / `xdg-open` open the directory in the OS
            // file manager.
            let opened = if cfg!(target_os = "macos") {
                detach("open", &[&dir_str], None)
            } else if cfg!(windows) {
                Command::new("explorer").arg(&dir_str).spawn().is_ok()
            } else {
                detach("xdg-open", &[&dir_str], None)
            };
            if opened {
                ShellOpResultWire {
                    ok: true,
                    error: None,
                }
            } else {
                ShellOpResultWire {
                    ok: false,
                    error: Some("Failed to open file manager".into()),
                }
            }
        }
        "terminal" => {
            if cfg!(target_os = "macos") {
                let ok = detach("open", &["-a", "Terminal", &dir_str], None);
                return result_or(ok, "Failed to launch Terminal");
            }
            if cfg!(windows) {
                let ok = detach("cmd", &["/c", "start", "", "cmd"], Some(dir));
                return result_or(ok, "Failed to launch cmd");
            }
            let working_dir = format!("--working-directory={dir_str}");
            let ok = detach("x-terminal-emulator", &[&working_dir], None)
                || detach("xdg-open", &[&dir_str], None);
            result_or(ok, "No terminal handler found")
        }
        "vscode" => result_or(
            launch_editor("code", "Visual Studio Code", dir),
            "Failed to launch VSCode",
        ),
        "zed" => result_or(launch_editor("zed", "Zed", dir), "Failed to launch Zed"),
        other => ShellOpResultWire {
            ok: false,
            error: Some(format!("Unknown target: {other}")),
        },
    }
}

fn result_or(ok: bool, error: &'static str) -> ShellOpResultWire {
    if ok {
        ShellOpResultWire {
            ok: true,
            error: None,
        }
    } else {
        ShellOpResultWire {
            ok: false,
            error: Some(error.into()),
        }
    }
}

/// `openInAppDetect`.
#[tauri::command]
pub fn open_in_app_detect() -> Vec<ExternalAppWire> {
    detect_apps()
}

/// Resolve a session's folder: worktree.path → session workspace path →
/// workspace-by-id → $HOME (the TS main.ts resolveSessionPath chain).
fn resolve_session_path(state: &AppState, session_id: Option<&str>) -> PathBuf {
    if let Some(session_id) = session_id.filter(|s| !s.is_empty()) {
        let db_path = state.sessions_db_path();
        if db_path.is_file() {
            if let Ok(store) = tide_store::sessions_v2::SessionsV2::open(&db_path) {
                if let Ok(Some(worktree)) = store.session_worktree_of(session_id) {
                    if let Some(path) = worktree.get("path").and_then(|v| v.as_str()) {
                        if Path::new(path).exists() {
                            return PathBuf::from(path);
                        }
                    }
                }
                if let Ok(Some(meta)) = store.session_meta_by_id(session_id) {
                    if Path::new(&meta.workspace_path).exists() {
                        return PathBuf::from(meta.workspace_path);
                    }
                }
            }
        }
        let by_id = state
            .read_config(|cfg| {
                cfg.workspaces
                    .iter()
                    .find(|ws| ws.id == session_id)
                    .map(|ws| ws.path.clone())
            })
            .ok()
            .flatten();
        if let Some(path) = by_id.filter(|p| Path::new(p).exists()) {
            return PathBuf::from(path);
        }
    }
    std::env::var("HOME")
        .map(PathBuf::from)
        .unwrap_or_else(|_| PathBuf::from("/"))
}

/// `openInAppOpen`.
#[tauri::command]
pub fn open_in_app_open(
    state: tauri::State<'_, AppState>,
    target: String,
    session_id: Option<String>,
) -> Result<ShellOpResultWire, CommandError> {
    let dir = resolve_session_path(&state, session_id.as_deref());
    Ok(open_in_target(&target, &dir))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn detect_lists_the_four_apps_with_icons_null() {
        let apps = open_in_app_detect();
        assert_eq!(apps.len(), 4);
        assert_eq!(apps[0].id, "finder");
        assert!(apps[0].available);
        assert!(apps[1].available, "terminal is always available");
        for app in &apps {
            assert!(app.icon_data_url.is_none());
        }
    }

    #[test]
    fn unknown_target_refuses() {
        let result = open_in_target("nope", Path::new("/tmp"));
        assert!(!result.ok);
        assert!(result.error.unwrap().contains("Unknown target"));
    }

    #[test]
    fn missing_path_refuses() {
        let result = open_in_target("finder", Path::new("/definitely/not/here/xyz"));
        assert!(!result.ok);
        assert!(result.error.unwrap().contains("Path does not exist"));
    }

    #[test]
    fn shell_op_wire_shape() {
        let ok = ShellOpResultWire {
            ok: true,
            error: None,
        };
        assert_eq!(
            serde_json::to_value(&ok).unwrap(),
            serde_json::json!({ "ok": true })
        );
        let err = ShellOpResultWire {
            ok: false,
            error: Some("nope".into()),
        };
        assert_eq!(
            serde_json::to_value(&err).unwrap(),
            serde_json::json!({ "ok": false, "error": "nope" })
        );
    }
}
